hls-exactprint-utils/src/Ide/TreeTransform.hs

```diff
@@ -0,0 +1,334 @@
+    (anns, val') <- annotate dflags $ maybeParensAST val
+    modifyAnnsT $ mappend anns
+    pure $
+        everywhere'
+            ( mkT $
+                \case
+                    (L src _ :: Located ast) | src == dst -> val'
+                    l -> l
+            )
+            a
+
+------------------------------------------------------------------------------
+
+graftWithM ::
+    forall ast m a.
+    (Fail.MonadFail m, Data a, ASTElement ast) =>
+    SrcSpan ->
+    (Located ast -> TransformT m (Maybe (Located ast))) ->
+    Graft m a
+graftWithM dst trans = Graft $ \dflags a -> do
+    everywhereM'
+        ( mkM $
+            \case
+                val@(L src _ :: Located ast)
+                    | src == dst -> do
+                        mval <- trans val
+                        case mval of
+                            Just val' -> do
+                                (anns, val'') <-
+                                    hoistTransform (either Fail.fail pure) $
+                                        annotate dflags $ maybeParensAST val'
+                                modifyAnnsT $ mappend anns
+                                pure val''
+                            Nothing -> pure val
+                l -> pure l
+        )
+        a
+
+graftWithSmallestM ::
+    forall ast m a.
+    (Fail.MonadFail m, Data a, ASTElement ast) =>
+    SrcSpan ->
+    (Located ast -> TransformT m (Maybe (Located ast))) ->
+    Graft m a
+graftWithSmallestM dst trans = Graft $ \dflags a -> do
+    everywhereM'
+        ( mkM $
+            \case
+                val@(L src _ :: Located ast)
+                    | dst `isSubspanOf` src -> do
+                        mval <- trans val
+                        case mval of
+                            Just val' -> do
+                                (anns, val'') <-
+                                    hoistTransform (either Fail.fail pure) $
+                                        annotate dflags $ maybeParensAST val'
+                                modifyAnnsT $ mappend anns
+                                pure val''
+                            Nothing -> pure val
+                l -> pure l
+        )
+        a
+
+graftDecls ::
+    forall a.
+    (HasDecls a) =>
+    SrcSpan ->
+    [LHsDecl GhcPs] ->
+    Graft (Either String) a
+graftDecls dst decs0 = Graft $ \dflags a -> do
+    decs <- forM decs0 $ \decl -> do
+        (anns, decl') <- annotateDecl dflags decl
+        modifyAnnsT $ mappend anns
+        pure decl'
+    let go [] = DL.empty
+        go (L src e : rest)
+            | src == dst = DL.fromList decs <> DL.fromList rest
+            | otherwise = DL.singleton (L src e) <> go rest
+    modifyDeclsT (pure . DL.toList . go) a
+
+graftDeclsWithM ::
+    forall a m.
+    (HasDecls a, Fail.MonadFail m) =>
+    SrcSpan ->
+    (LHsDecl GhcPs -> TransformT m (Maybe [LHsDecl GhcPs])) ->
+    Graft m a
+graftDeclsWithM dst toDecls = Graft $ \dflags a -> do
```

Comment:
@konn I also don't understand what this function is doing. Mind walking me through it? Especially what `go` does.

Reply:
The intended usage of `graftDeclsWithM` is to apply a one-to-many transformation on declaration lists.
This transformation is needed when I implemented the expansion of declaration splices, as a single declaration splice can be expanded into multiple declarations.

`go` is just doing something like `concatMapM` (or `fmap getAp . foldMap . fmap Ap`), based on span information: it applies transformation when the span matched, replace a single declaration with the generated multiple declarations if the transformation succeeded.
At the time I wrote it, I couldn't find any one-to-many grafting function in ghc-exactprint, so I wrote this by hand.